hns_parse_ptr_reply.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.\"
.TH HNS_PARSE_PTR_REPLY 3 "25 July 1998"
.SH NAME
hns_parse_ptr_reply \- Parse a reply to a DNS query of type PTR into a hostent
.SH SYNOPSIS
.nf
.B #include <hns.h>
.PP
.B int hns_parse_ptr_reply(const unsigned char *\fIabuf\fP, int \fIalen\fP,
.B 	const void *\fIaddr\fP, int \fIaddrlen\fP, int \fIfamily\fP,
.B	struct hostent **\fIhost\fP);
.fi
.SH DESCRIPTION
The
.B hns_parse_ptr_reply
function parses the response to a query of type PTR into a
.BR "struct hostent" .
The parameters
.I abuf
and
.I alen
give the contents of the response.  The parameters
.IR addr ,
.IR addrlen ,
and
.I family
specify which address was queried for; they are not used to verify the
response, merely used to fill in the address of the
.BR "struct hostent" .
The resulting
.B struct hostent
is stored in allocated memory and a pointer to it stored into the
variable pointed to by
.IR host .
It is the caller's responsibility to free the resulting host structure
using
.BR hns_free_hostent (3)
when it is no longer needed.
.SH RETURN VALUES
.B hns_parse_ptr_reply
can return any of the following values:
.TP 15
.B HNS_SUCCESS
The response was successfully parsed.
.TP 15
.B HNS_EBADRESP
The response was malformatted.
.TP 15
.B HNS_ENODATA
The response did not contain an answer to the query.
.TP 15
.B HNS_ENOMEM
Memory was exhausted.
.SH SEE ALSO
.BR hns_gethostbyaddr (3),
.BR hns_free_hostent (3)
.SH AUTHOR
Greg Hudson, MIT Information Systems
.br
Copyright 1998 by the Massachusetts Institute of Technology.


NEWS
Major changes since:
* see the CHANGES file

Major changes in release 1.1.1:
* hns should now compile as C++ code (no longer uses reserved word
  "class").
* Added SRV support to hdig test program.
* Fixed a few error handling bugs in query processing.

Major changes in release 1.1.0:
* Added hns_free_string() function so that memory can be freed in the
  same layer as it is allocated, a desirable feature in some
  environments.
* A few of the hns_dns.h macros are fixed to use the proper bitwise
  operator.
* Fixed a couple of fenceposts fixed in hns_expand_name()'s
  bounds-checking.
* In process_timeouts(), extract query->next before calling
  next_server() and possibly freeing the query structure.
* Casted arguments to ctype macros casted to unsigned char, since not
  all char values are valid inputs to those macros according to ANSI.


hns_parse_ptr_reply.c

/* Copyright 1998 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

#include "hns_setup.h"

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#ifdef HAVE_NETDB_H
#  include <netdb.h>
#endif
#ifdef HAVE_ARPA_NAMESER_H
#  include <arpa/nameser.h>
#else
#  include "nameser.h"
#endif
#ifdef HAVE_ARPA_NAMESER_COMPAT_H
#  include <arpa/nameser_compat.h>
#endif

#ifdef HAVE_STRINGS_H
#  include <strings.h>
#endif

#include "hns.h"
#include "hns_dns.h"
#include "hns_nowarn.h"
#include "hns_private.h"

int hns_parse_ptr_reply(const unsigned char *abuf, int alen, const void *addr,
                         int addrlen, int family, struct hostent **host)
{
  unsigned int qdcount, ancount;
  int status, i, rr_type, rr_class, rr_len;
  long len;
  const unsigned char *aptr;
  char *ptrname, *hostname, *rr_name, *rr_data;
  struct hostent *hostent;
  int aliascnt = 0;
  int alias_alloc = 8;
  char ** aliases;

  /* Set *host to NULL for all failure cases. */
  *host = NULL;

  /* Give up if abuf doesn't have room for a header. */
  if (alen < HFIXEDSZ)
    return HNS_EBADRESP;

  /* Fetch the question and answer count from the header. */
  qdcount = DNS_HEADER_QDCOUNT(abuf);
  ancount = DNS_HEADER_ANCOUNT(abuf);
  if (qdcount != 1)
    return HNS_EBADRESP;

  /* Expand the name from the question, and skip past the question. */
  aptr = abuf + HFIXEDSZ;
  status = hns__expand_name_for_response(aptr, abuf, alen, &ptrname, &len);
  if (status != HNS_SUCCESS)
    return status;
  if (aptr + len + QFIXEDSZ > abuf + alen)
    {
      hns_free(ptrname);
      return HNS_EBADRESP;
    }
  aptr += len + QFIXEDSZ;

  /* Examine each answer resource record (RR) in turn. */
  hostname = NULL;
  aliases = hns_malloc(alias_alloc * sizeof(char *));
  if (!aliases)
    {
      hns_free(ptrname);
      return HNS_ENOMEM;
    }
  for (i = 0; i < (int)ancount; i++)
    {
      /* Decode the RR up to the data field. */
      status = hns__expand_name_for_response(aptr, abuf, alen, &rr_name, &len);
      if (status != HNS_SUCCESS)
        break;
      aptr += len;
      if (aptr + RRFIXEDSZ > abuf + alen)
        {
          hns_free(rr_name);
          status = HNS_EBADRESP;
          break;
        }
      rr_type = DNS_RR_TYPE(aptr);
      rr_class = DNS_RR_CLASS(aptr);
      rr_len = DNS_RR_LEN(aptr);
      aptr += RRFIXEDSZ;
      if (aptr + rr_len > abuf + alen)
        {
          hns_free(rr_name);
          status = HNS_EBADRESP;
          break;
        }

      if (rr_class == C_IN && rr_type == T_PTR
          && strcasecmp(rr_name, ptrname) == 0)
        {
          /* Decode the RR data and set hostname to it. */
          status = hns__expand_name_for_response(aptr, abuf, alen, &rr_data,
                                                  &len);
          if (status != HNS_SUCCESS)
            {
              hns_free(rr_name);
              break;
            }
          if (hostname)
            hns_free(hostname);
          hostname = rr_data;
          aliases[aliascnt] = hns_malloc((strlen(rr_data)+1) * sizeof(char));
          if (!aliases[aliascnt])
            {
              hns_free(rr_name);
              status = HNS_ENOMEM;
              break;
            }
          strncpy(aliases[aliascnt], rr_data, strlen(rr_data)+1);
          aliascnt++;
          if (aliascnt >= alias_alloc) {
            char **ptr;
            alias_alloc *= 2;
            ptr = hns_realloc(aliases, alias_alloc * sizeof(char *));
            if(!ptr) {
              hns_free(rr_name);
              status = HNS_ENOMEM;
              break;
            }
            aliases = ptr;
          }
        }

      if (rr_class == C_IN && rr_type == T_CNAME)
        {
          /* Decode the RR data and replace ptrname with it. */
          status = hns__expand_name_for_response(aptr, abuf, alen, &rr_data,
                                                  &len);
          if (status != HNS_SUCCESS)
            {
              hns_free(rr_name);
              break;
            }
          hns_free(ptrname);
          ptrname = rr_data;
        }

      hns_free(rr_name);
      aptr += rr_len;
      if (aptr > abuf + alen)
        {  /* LCOV_EXCL_START: already checked above */
          status = HNS_EBADRESP;
          break;
        }  /* LCOV_EXCL_STOP */
    }

  if (status == HNS_SUCCESS && !hostname)
    status = HNS_ENODATA;
  if (status == HNS_SUCCESS)
    {
      /* We got our answer.  Allocate memory to build the host entry. */
      hostent = hns_malloc(sizeof(struct hostent));
      if (hostent)
        {
          hostent->h_addr_list = hns_malloc(2 * sizeof(char *));
          if (hostent->h_addr_list)
            {
              hostent->h_addr_list[0] = hns_malloc(addrlen);
              if (hostent->h_addr_list[0])
                {
                  hostent->h_aliases = hns_malloc((aliascnt+1) * sizeof (char *));
                  if (hostent->h_aliases)
                    {
                      /* Fill in the hostent and return successfully. */
                      hostent->h_name = hostname;
                      for (i=0 ; i<aliascnt ; i++)
                        hostent->h_aliases[i] = aliases[i];
                      hostent->h_aliases[aliascnt] = NULL;
                      hostent->h_addrtype = hnsx_sitoss(family);
                      hostent->h_length = hnsx_sitoss(addrlen);
                      memcpy(hostent->h_addr_list[0], addr, addrlen);
                      hostent->h_addr_list[1] = NULL;
                      *host = hostent;
                      hns_free(aliases);
                      hns_free(ptrname);
                      return HNS_SUCCESS;
                    }
                  hns_free(hostent->h_addr_list[0]);
                }
              hns_free(hostent->h_addr_list);
            }
          hns_free(hostent);
        }
      status = HNS_ENOMEM;
    }
  for (i=0 ; i<aliascnt ; i++)
    if (aliases[i])
      hns_free(aliases[i]);
  hns_free(aliases);
  if (hostname)
    hns_free(hostname);
  hns_free(ptrname);
  return status;
}
